### PR TITLE
Enumerable module

### PR DESCRIPTION
**Pull request by sz16900**
# Completed [Project # 9 Bubble Sort](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks?ref=lnav) in partial requirements of the [Microverse Curriculum](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/57421588/).

---

### Contributors:
- Seth Zea

--- 

##### General Requirements

- [x] Add a descriptive README file to your repository.
- [x] Use meaningful names for your Pull Requests and add a short summary of its content.
- [x] Use descriptive commit messages.
- [x] Do not commit files that are not related to the project. Instead, add them to .gitignore file.
- [x] Do not commit inline comments. 
- [x] Use stickler for checks.

##### Other Requirements
- [x] Keep your code clean.
- [x] Do not commit debugging instructions

##### Project Specific
- [x] Remember that you should use the Git Flow for the first time. Pay attention to the “Remember to” and “Code review” sections.
- [x] Take a look at Ruby docs about Enumerable class https://ruby-doc.org/core-2.6.5/Enumerable.html

# Project Status :

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice progress :+1: @sz16900 
You are almost there. We need to fix a few things before we can merge this one.
Check my comments on how to fix them.

### Martin Afani

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      unless block_given?
+        return false unless x
+
+        return true
+      end
+
+      return false unless yield x
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

Sample cases:
```ruby
let(:true_array) { [1, true, 'hi', []] }
let(:false_array) { [1, false, 'hi', []] }

expect(true_array.my_all?).to be true_array.all?
expect(false_array.my_all?).to be false_array.all?
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    my_each do |x|
+      return true if reg_exp(pattern, x) || num_exp(pattern, x)
+
+      unless block_given?
+        return true if x
+      end
+      return true if yield x
+    end
+    false
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

Sample testcase:
```ruby
let(:true_array) { [nil, false, true, []] }
let(:false_array) { [nil, false, nil, false] }

expect(true_array.my_any?).to be true_array.any?
expect(false_array.my_any?).to be false_array.any?
```
Should return `true` if `my_any?` works well

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+      unless block_given?
+        return false if x
+
+        return true
+      end
+      return false if yield x
+    end
+    true
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

Sample test case:
```ruby
let(:true_array) { [nil, false, true, []] }
let(:false_array) { [nil, false, nil, false] }

expect(false_array.my_none?).to be true
expect(true_array.my_none?).to be false
```
Should return `true` and `false`

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      end
+      unless block_given?
+        return repetitions += 1 if (yield x) == x
+
+        return repetitions
+      end
+    end
+    repetitions += 1 while repetitions < size
+    repetitions
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns the number of items in enum through enumeration
- [x] counts the number of items in enum that are equal to item if an argument is given
- [x] counts the number of elements yielding a true value if a block is given

Sample test case:
```ruby
let(:array) { Array.new(100) { rand(0...9) } }
expect(array.my_count).to eq array.count
```
Should return `true`

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    arr = []
+    each do |x|
+      arr << if block_given?
+               (yield x)
+             else
+               proc.call(x)
+             end
+    end
+    arr
+  end
```
Comment (whiz25) [inline enumerables.rb]: - [ ] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given
Sample test case:
```ruby
let(:array) { Array.new(100) { rand(0...9) } }
expect(array.my_map).to be_an(Enumerator)
```
Should return `true`

**Review by Forison — APPROVED** [review]
## Status: Approved💚 
# 🙌 Zea,
Good job on implementing the changes suggested earlier, go ahead and merge.
## Proceed to the next task👌.
### Happy coding💻
### Addo Forison

**Pull request merged**